app_dart/lib/src/request_handlers/dart_internal_subscription.dart

```diff
@@ -2,31 +2,126 @@
 import '../service/logging.dart';
 
+/// TODO(drewroengoogle): Make this subscription generic so we can accept more
+/// than just dart-internal builds.
+///
 /// An endpoint for listening to build updates for dart-internal builds and
 /// saving the results to the datastore.
 ///
 /// The PubSub subscription is set up here:
 /// https://console.cloud.google.com/cloudpubsub/subscription/detail/dart-internal-build-results-sub?project=flutter-dashboard
 @immutable
 class DartInternalSubscription extends SubscriptionHandler {
-  /// Creates an endpoint for listening to dart-internal build results.
+  /// Creates an endpoint for listening for dart-internal build results.
+  /// The message should contain a single buildbucket id
   const DartInternalSubscription({
     required super.cache,
     required super.config,
     super.authProvider,
+    required this.buildBucketClient,
+    @visibleForTesting this.datastoreProvider = DatastoreService.defaultProvider,
+    this.retryOptions = Config.buildbucketRetry,
   }) : super(subscriptionName: 'dart-internal-build-results-sub');
 
+  final BuildBucketClient buildBucketClient;
+  final DatastoreServiceProvider datastoreProvider;
+  final RetryOptions retryOptions;
+
   @override
   Future<Body> post() async {
-    log.fine('Consumed message: ${message.data}');
+    final DatastoreService datastore = datastoreProvider(config.db);
+
+    log.info("Converting message data to non nullable int");
+
+    final int buildbucketId = int.parse(json.decode(message.data.toString())['buildbucket_id']);
+
+    log.info("Creating build request object");
+    final GetBuildRequest request = GetBuildRequest(id: buildbucketId.toString());
+
+    log.info(
+      "Calling buildbucket api to get build data for build $buildbucketId",
+    );
+    final Build build = await retryOptions.retry(
+      () async {
+        return _getBuildFromBuildbucket(request);
+      },
+      retryIf: (Exception e) => e is UnfinishedBuildException,
+    );
+
+    log.info("Checking for existing task in datastore");
+    final Task? existingTask = await _getExistingTaskFromDatastore(build, datastore);
+
+    late Task taskToInsert;
+    if (existingTask != null) {
+      log.info("Updating Task from existing Task");
+      existingTask.buildNumber = build.number;
+      existingTask.buildNumberList = '${build.number},${existingTask.buildNumberList}';
+      taskToInsert = existingTask;
+    } else {
+      log.info("Creating Task from Buildbucket result");
+      taskToInsert = await Task.fromBuildbucketResult(build, datastore);
+    }
+
+    log.info("Inserting Task into the datastore: ${taskToInsert.toString()}");
+    await datastore.insert(<Task>[taskToInsert]);
+
+    return Body.forJson(taskToInsert.toString());
+  }
+
+  Future<Build> _getBuildFromBuildbucket(
+    GetBuildRequest request,
+  ) async {
+    final Build build = await buildBucketClient.getBuild(request);
+
+    if (build.status == Status.started) {
```

Comment:
What are the possible states here? I would assume more than one terminal state? Might be better to look for those instead if you are testing for finished.

Reply:
Good point. I updated it to check for success/failure/canceled instead.